File: sdk/java/src/main/java/com/anyvali/schemas/StringSchema.java
package com.anyvali.schemas;

import com.anyvali.IssueCodes;
import com.anyvali.Schema;
import com.anyvali.ValidationContext;
import com.anyvali.format.FormatValidators;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Schema for string values with optional constraints.
 */
public class StringSchema extends Schema<String> {
    private Integer minLength;
    private Integer maxLength;
    private String pattern;
    private String startsWith;
    private String endsWith;
    private String includes;
    private String format;

    public StringSchema() {}

    private StringSchema(StringSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
        this.metadata = other.metadata;
        this.minLength = other.minLength;
        this.maxLength = other.maxLength;
        this.pattern = other.pattern;
        this.startsWith = other.startsWith;
        this.endsWith = other.endsWith;
        this.includes = other.includes;
        this.format = other.format;
    }

    public StringSchema minLength(int n) {
        var copy = new StringSchema(this);
        copy.minLength = n;
        return copy;
    }

    public StringSchema maxLength(int n) {
        var copy = new StringSchema(this);
        copy.maxLength = n;
        return copy;
    }

    public StringSchema pattern(String p) {
        var copy = new StringSchema(this);
        copy.pattern = p;
        return copy;
    }

    public StringSchema startsWith(String s) {
        var copy = new StringSchema(this);
        copy.startsWith = s;
        return copy;
    }

    public StringSchema endsWith(String s) {
        var copy = new StringSchema(this);
        copy.endsWith = s;
        return copy;
    }

    public StringSchema includes(String s) {
        var copy = new StringSchema(this);
        copy.includes = s;
        return copy;
    }

    public StringSchema format(String f) {
        var copy = new StringSchema(this);
        copy.format = f;
        return copy;
    }

    @Override
    protected Object validate(Object input, ValidationContext ctx) {
        if (!(input instanceof String s)) {
            String received = anyvaliTypeName(input);
            ctx.addIssue(IssueCodes.INVALID_TYPE,
                    "Expected string, received " + received, "string", received);
            return null;
        }

        int length = s.codePointCount(0, s.length());

        if (minLength != null && length < minLength) {
            ctx.addIssue(IssueCodes.TOO_SMALL,
                    "String must have at least " + minLength + " character(s)",
                    minLength, length);
        }

        if (maxLength != null && length > maxLength) {
            ctx.addIssue(IssueCodes.TOO_LARGE,
                    "String must have at most " + maxLength + " character(s)",
                    maxLength, length);
        }

        if (pattern != null) {
            try {
                if (!Pattern.compile(toEcmaAnchors(pattern)).matcher(s).find()) {
                    ctx.addIssue(IssueCodes.INVALID_STRING,
                            "String does not match pattern '" + pattern + "'",
                            pattern, s);
                }
            } catch (Exception e) {
                // Invalid regex pattern - treat as validation failure
                ctx.addIssue(IssueCodes.INVALID_STRING,
                        "Invalid regex pattern: " + pattern,
                        pattern, s);
            }
        }

        if (startsWith != null && !s.startsWith(startsWith)) {
            ctx.addIssue(IssueCodes.INVALID_STRING,
                    "String must start with '" + startsWith + "'",
                    startsWith, s);
        }

        if (endsWith != null && !s.endsWith(endsWith)) {
            ctx.addIssue(IssueCodes.INVALID_STRING,
                    "String must end with '" + endsWith + "'",
                    endsWith, s);
        }

        if (includes != null && !s.contains(includes)) {
            ctx.addIssue(IssueCodes.INVALID_STRING,
                    "String must include '" + includes + "'",
                    includes, s);
        }

        if (format != null && !FormatValidators.validate(format, s)) {
            ctx.addIssue(IssueCodes.INVALID_STRING,
                    "Invalid " + format, format, s);
        }

        return s;
    }

    @Override
    protected Map<String, Object> toNode() {
        var node = new LinkedHashMap<String, Object>();
        node.put("kind", "string");
        if (minLength != null) node.put("minLength", minLength);
        if (maxLength != null) node.put("maxLength", maxLength);
        if (pattern != null) node.put("pattern", pattern);
        if (startsWith != null) node.put("startsWith", startsWith);
        if (endsWith != null) node.put("endsWith", endsWith);
        if (includes != null) node.put("includes", includes);
        if (format != null) node.put("format", format);
        return addCommonNodeFields(node);
    }

    /**
     * Rewrite ECMA-262 anchors to Java's absolute anchors. In ECMA without the
     * multiline flag, "^"/"$" match only the start/end of the whole string.
     * Java's "$" also matches just before a final line terminator, so an
     * anchored whitelist like ^[a-z]+$ would accept "abc\n" -- a
     * newline-injection bypass that diverges from the JS reference. Translate
     * unescaped, top-level "^" -> "\A" and "$" -> "\z" (absolute end). Anchors
     * inside character classes and escaped "\^"/"\$" are left untouched.
     */
    private static String toEcmaAnchors(String p) {
        var sb = new StringBuilder(p.length() + 4);
        boolean escaped = false;
        boolean inClass = false;
        for (int i = 0; i < p.length(); i++) {
            char ch = p.charAt(i);
            if (escaped) { sb.append(ch); escaped = false; }
            else if (ch == '\\') { sb.append(ch); escaped = true; }
            else if (ch == '[') { inClass = true; sb.append(ch); }
            else if (ch == ']' && inClass) { inClass = false; sb.append(ch); }
            else if (ch == '^' && !inClass) { sb.append("\\A"); }
            else if (ch == '$' && !inClass) { sb.append("\\z"); }
            else { sb.append(ch); }
        }
        return sb.toString();
    }

    @Override
    protected Schema<String> copy() {
        return new StringSchema(this);
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/StringSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import com.anyvali.format.FormatValidators
import com.anyvali.parse.Coercion
import kotlinx.serialization.json.*

data class StringSchema(
    val minLength: Int? = null,
    val maxLength: Int? = null,
    val pattern: String? = null,
    val startsWith: String? = null,
    val endsWith: String? = null,
    val includes: String? = null,
    val format: String? = null,
    val defaultValue: Any? = UNSET,
    val coerce: List<String> = emptyList()
) : Schema<String>() {
    override val kind: String = "string"

    fun minLength(n: Int) = copy(minLength = n)
    fun maxLength(n: Int) = copy(maxLength = n)
    fun pattern(p: String) = copy(pattern = p)
    fun startsWith(s: String) = copy(startsWith = s)
    fun endsWith(s: String) = copy(endsWith = s)
    fun includes(s: String) = copy(includes = s)
    fun format(f: String) = copy(format = f)
    fun default(v: String) = copy(defaultValue = v)
    fun coerce(vararg c: String) = copy(coerce = c.toList())

    fun describe(description: String, opts: DescribeOptions? = null): StringSchema {
        applyDescribe(description, opts)
        return this
    }

    fun metadata(meta: Map<String, Any?>, replace: Boolean = false): StringSchema {
        applyMetadata(meta, replace)
        return this
    }

    override fun safeParseWithContext(input: Any?, ctx: ValidationContext): ParseResult<String> {
        var value = input

        // Apply coercions
        if (coerce.isNotEmpty() && value is String) {
            for (c in coerce) {
                value = when (c) {
                    "trim" -> (value as String).trim()
                    "lower" -> (value as String).lowercase()
                    "upper" -> (value as String).uppercase()
                    else -> value
                }
            }
        }

        val issues = validateValue(value, ctx)
        return if (issues.isEmpty()) {
            ParseResult.Success(value as String)
        } else {
            ParseResult.Failure(issues)
        }
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        if (value !is String) {
            return listOf(
                ValidationIssue(
                    code = IssueCodes.INVALID_TYPE,
                    path = ctx.path,
                    expected = "string",
                    received = getJsonTypeName(value)
                )
            )
        }
        val issues = mutableListOf<ValidationIssue>()
        val length = value.codePointCount(0, value.length)

        minLength?.let {
            if (length < it) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.TOO_SMALL,
                        path = ctx.path,
                        expected = it.toString(),
                        received = length.toString()
                    )
                )
            }
        }
        maxLength?.let {
            if (length > it) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.TOO_LARGE,
                        path = ctx.path,
                        expected = it.toString(),
                        received = length.toString()
                    )
                )
            }
        }
        pattern?.let {
            try {
                if (!Regex(toEcmaAnchors(it)).containsMatchIn(value)) {
                    issues.add(
                        ValidationIssue(
                            code = IssueCodes.INVALID_STRING,
                            path = ctx.path,
                            expected = it,
                            received = value
                        )
                    )
                }
            } catch (_: Exception) {
                // Invalid regex pattern - treat as validation failure
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.INVALID_STRING,
                        path = ctx.path,
                        expected = it,
                        received = value
                    )
                )
            }
        }
        startsWith?.let {
            if (!value.startsWith(it)) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.INVALID_STRING,
                        path = ctx.path,
                        expected = it,
                        received = value
                    )
                )
            }
        }
        endsWith?.let {
            if (!value.endsWith(it)) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.INVALID_STRING,
                        path = ctx.path,
                        expected = it,
                        received = value
                    )
                )
            }
        }
        includes?.let {
            if (!value.contains(it)) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.INVALID_STRING,
                        path = ctx.path,
                        expected = it,
                        received = value
                    )
                )
            }
        }
        format?.let {
            if (!FormatValidators.validate(it, value)) {
                issues.add(
                    ValidationIssue(
                        code = IssueCodes.INVALID_STRING,
                        path = ctx.path,
                        expected = it,
                        received = value
                    )
                )
            }
        }

        return issues
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("string"))
        minLength?.let { put("minLength", JsonPrimitive(it)) }
        maxLength?.let { put("maxLength", JsonPrimitive(it)) }
        pattern?.let { put("pattern", JsonPrimitive(it)) }
        startsWith?.let { put("startsWith", JsonPrimitive(it)) }
        endsWith?.let { put("endsWith", JsonPrimitive(it)) }
        includes?.let { put("includes", JsonPrimitive(it)) }
        format?.let { put("format", JsonPrimitive(it)) }
        if (defaultValue !== UNSET) put("default", JsonPrimitive(defaultValue as? String))
        if (coerce.isNotEmpty()) {
            if (coerce.size == 1) put("coerce", JsonPrimitive(coerce[0]))
            else put("coerce", JsonArray(coerce.map { JsonPrimitive(it) }))
        }
        addMetadataToNode(this)
    }

    companion object {
        val UNSET = Any()

        /**
         * Rewrite ECMA-262 anchors to absolute anchors. In ECMA without the
         * multiline flag, "^"/"$" match only the start/end of the whole string.
         * Kotlin/Java "$" also matches just before a final line terminator, so
         * an anchored whitelist like ^[a-z]+$ would accept "abc\n" -- a
         * newline-injection bypass that diverges from the JS reference.
         * Translate unescaped, top-level "^" -> "\A" and "$" -> "\z". Anchors
         * inside character classes and escaped "\^"/"\$" are left untouched.
         */
        fun toEcmaAnchors(p: String): String {
            val sb = StringBuilder(p.length + 4)
            var escaped = false
            var inClass = false
            for (ch in p) {
                when {
                    escaped -> { sb.append(ch); escaped = false }
                    ch == '\\' -> { sb.append(ch); escaped = true }
                    ch == '[' -> { inClass = true; sb.append(ch) }
                    ch == ']' && inClass -> { inClass = false; sb.append(ch) }
                    ch == '^' && !inClass -> sb.append("\\A")
                    ch == '$' && !inClass -> sb.append("\\z")
                    else -> sb.append(ch)
                }
            }
            return sb.toString()
        }
    }
}
